theano/sandbox/mkl/__init__.py

```diff
@@ -0,0 +1,254 @@
+
+    if config.mkl.nn.enabled == "False":
+        mkl_available.msg = "MKL is disabled by the 'mkl.nn.enabled' setting."
+        return False
+
+    return True
+
+
+def mkl_available():
+    # Check this status every time which allows user
+    # to change their setting in different stage of model
+    if _dnn_is_enabled_by_user() is not True:
+        return False
+
+    if _dnn_is_able_to_link() is not True:
+        if config.mkl.nn.enabled == "True":
+            raise RuntimeError("You enabled MKL, but we aren't able "
+                               "to use it, %s" % mkl_available.msg)
+        return False
+
+    return True
+
+mkl_available.msg = None
+
+
+if mkl_available():
+    mkl_path = os.path.abspath(os.path.split(__file__)[0])
+    mkl_ndarray_loc = os.path.join(config.compiledir, 'mkl_ndarray')
+    mkl_ndarray_so = os.path.join(
+        mkl_ndarray_loc, 'mkl_ndarray.' + get_lib_extension())
+    libmkl_ndarray_so = os.path.join(
+        mkl_ndarray_loc, 'libmkl_ndarray.' + get_lib_extension())
+
+    def try_import_mkl_ndarray():
+        mkl_files = ('mkl_ndarray.c', 'mkl_ndarray.h')
+
+        stat_times = [os.stat(os.path.join(mkl_path, mkl_file))[stat.ST_MTIME]
+                      for mkl_file in mkl_files]
+        date = max(stat_times)
+        if os.path.exists(mkl_ndarray_so):
+            if date >= os.stat(mkl_ndarray_so)[stat.ST_MTIME]:
+                return False
+
+        try:
+            sys.path[0:0] = [config.compiledir]
+            import mkl_ndarray.mkl_ndarray  # noqa
+            del sys.path[0]
+        except ImportError:
+            return False
+        return True
+
+    compile_mkl_ndarray = not try_import_mkl_ndarray()
+
+    if compile_mkl_ndarray:
+        get_lock()
+        try:
+            if not try_import_mkl_ndarray():
+                try:
+                    code = open(os.path.join(mkl_path, 'mkl_ndarray.c')).read()
+                    if not os.path.exists(mkl_ndarray_loc):
+                        os.makedirs(mkl_ndarray_loc)
+                    compiler = GCC_compiler()
+                    preargs = compiler.compile_args()
+
+                    if 'mklml_intel' in config.blas.ldflags:
+                        mkl_lib = 'mklml_intel'
+                    elif 'mklml_gnu' in config.blas.ldflags:
+                        mkl_lib = 'mklml_gnu'
+                    else:
+                        mkl_lib = 'mkl_rt'
+                    compiler.compile_str('mkl_ndarray',
+                                         code,
+                                         location=mkl_ndarray_loc,
+                                         include_dirs=[mkl_path],
+                                         libs=[mkl_lib],
+                                         preargs=preargs,)
+
+                    from mkl_ndarray.mkl_ndarray import *  # noqa
+                except Exception as e:
+                    _logger.error('Failed to compile mkl_ndarray.c: %s', str(e))
```

Comment:
All that remind me that we want to refactor/ease compiling other lib with Theano. It probably won't be done shortly, so don't change the PR, but we should do it at some point.